x/examples/measure/main.go

```diff
@@ -0,0 +1,286 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"log/slog"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport/tls"
+	"github.com/Jigsaw-Code/outline-sdk/x/configurl"
+	"github.com/goccy/go-yaml"
+	"github.com/lmittmann/tint"
+	"golang.org/x/term"
+)
+
+func init() {
+	flag.Usage = func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "Usage: %s [flags...] <url>\n", path.Base(os.Args[0]))
+		flag.PrintDefaults()
+	}
+}
+
+type MeasurementConfig struct {
+	NumAttempts *int     `yaml:"num_attempts,omitempty"`
+	Proto       string   `yaml:"proto,omitempty"`
+	Domains     []string `yaml:"domains,omitempty"`
+	ISPProxies  []string `yaml:"isp_proxies,omitempty"`
+	Strategies  []string `yaml:"strategies,omitempty"`
+}
+
+type Measurement struct {
+	Time         time.Time
+	Domain       string
+	ISP          string
+	Country      string
+	Strategy     string
+	Attempt      int
+	Img          string
+	ErrorOp      string
+	ErrorMessage string
+}
+
+type ISP struct {
+	Transport   string
+	Name        string
+	CountryCode string
+}
+
+type ISPInfo struct {
+	Data struct {
+		IP          string
+		ISP         string
+		Carrier     string
+		City        string
+		Region      string
+		CountryName string `json:"country_name,omitempty"`
+		CountryCode string `json:"country_code,omitempty"`
+	}
+}
+
+var transportToDialer = configurl.NewDefaultProviders()
+
+func makeErrorMsg(err error, domain string) string {
+	return strings.ReplaceAll(err.Error(), domain, "${DOMAIN}")
+}
+
+func newISP(ispProxy string) ISP {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dialer, err := transportToDialer.NewStreamDialer(ctx, ispProxy)
+	if err != nil {
+		slog.Error("Could not create ISP dialer", "error", err)
+		os.Exit(1)
+	}
+	dialContext := func(ctx context.Context, network, addr string) (net.Conn, error) {
+		if !strings.HasPrefix(network, "tcp") {
+			return nil, fmt.Errorf("protocol not supported: %v", network)
+		}
+		return dialer.DialStream(ctx, addr)
+	}
+	httpClient := &http.Client{
+		Transport: &http.Transport{DialContext: dialContext},
+		Timeout:   time.Duration(10) * time.Second,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	req, err := http.NewRequest("GET", "https://checker.soax.com/api/ipinfo", nil)
```

Comment:
In the future we should change to DNS requests (e.g. whoami at Akamai) to some service, or using STUN to get the IP